### PR TITLE
 KEYCLOAK-14317 [sso74-dev] Stop using service ca from service account token

### PR DESCRIPTION
**Pull request by marun**
This change is required by [Bugzilla 1813894](https://bugzilla.redhat.com/show_bug.cgi?id=1813894). Inclusion of the service ca in token configmaps is discontinued in 4.5.

### CODE REVIEW

File: templates/openj9/sso74-openj9-x509-https.json [hunk 1]
```diff
@@ -369,6 +388,12 @@
                                 "secret": {
                                     "secretName": "sso-x509-jgroups-secret"
                                 }
+                            },
+                            {
+                                "name": "service-ca",
+                                "configMap": {
+                                    "configMapName": "${APPLICATION_NAME}-service-ca"
```
Comment (iankko) [inline templates/openj9/sso74-openj9-x509-https.json]: @marun Replace `"configMapName"` with `"name"`, IOWs do:
```
[iankko@localhost tmp]$ diff -pruN sso74-x509-https.json.orig sso74-x509-https.json
--- sso74-x509-https.json.orig	2020-05-22 14:22:20.564266004 +0200
+++ sso74-x509-https.json	2020-05-22 14:25:35.693080510 +0200
@@ -392,7 +392,7 @@
                             {
                                 "name": "service-ca",
                                 "configMap": {
-                                    "configMapName": "${APPLICATION_NAME}-service-ca"
+                                    "name": "${APPLICATION_NAME}-service-ca"
                                 }
                             }
                         ]
```

In the current version the template, when calling `oc new-app --template=...` returns:
```
[iankko@localhost tmp]$ oc new-app --template=sso74-x509-https
--> Deploying template "iankko-inject-ca-bundle/sso74-x509-https" to project iankko-inject-ca-bundle

     Red Hat Single Sign-On 7.4 on OpenJDK (Ephemeral)
     ---------
     An example application based on RH-SSO 7.4 on OpenJDK image. For more information about using this template, see https://github.com/jboss-container-images/redhat-sso-7-openshift-image/tree/sso74-dev/docs.

     A new RH-SSO service has been created in your project. The admin username/password for accessing the master realm via the RH-SSO console is X86pvCyd/dMofOUHii7yIgsikM76ftX8ABG1HnUCf. The HTTPS keystore used for serving secure content, the JGroups keystore used for securing JGroups communications, and server truststore used for securing RH-SSO requests were automatically created via OpenShift's service serving x509 certificate secrets.

     * With parameters:
        * Application Name=sso
        * Custom RH-SSO Server Hostname=
        * JGroups Cluster Password=Gfire5WQssryMmI5LHG65VCpnswntTpy # generated
        * Datasource Minimum Pool Size=
        * Datasource Maximum Pool Size=
        * Datasource Transaction Isolation=
        * ImageStream Namespace=openshift
        * RH-SSO Administrator Username=X86pvCyd # generated
        * RH-SSO Administrator Password=dMofOUHii7yIgsikM76ftX8ABG1HnUCf # generated
        * RH-SSO Realm=
        * RH-SSO Service Username=
        * RH-SSO Service Password=
        * Container Memory Limit=1Gi

--> Creating resources ...
    configmap "sso-service-ca" created
    service "sso" created
    service "sso-ping" created
    route.route.openshift.io "sso" created
    error: DeploymentConfig.apps.openshift.io "sso" is invalid: [spec.template.spec.volumes[2].configMap.name: Required value, spec.template.spec.containers[0].volumeMounts[2].name: Not found: "service-ca"]
--> Failed
```

In the replaced form, it works correctly:
```
[iankko@localhost tmp]$ oc new-app --template=sso74-x509-https
--> Deploying template "iankko-inject-ca-bundle/sso74-x509-https" to project iankko-inject-ca-bundle

     Red Hat Single Sign-On 7.4 on OpenJDK (Ephemeral)
     ---------
     An example application based on RH-SSO 7.4 on OpenJDK image. For more information about using this template, see https://github.com/jboss-container-images/redhat-sso-7-openshift-image/tree/sso74-dev/docs.

     A new RH-SSO service has been created in your project. The admin username/password for accessing the master realm via the RH-SSO console is MGOBwlID/OkqS07eehu1cvdOmAPr3s6EJrqBhOMqq. The HTTPS keystore used for serving secure content, the JGroups keystore used for securing JGroups communications, and server truststore used for securing RH-SSO requests were automatically created via OpenShift's service serving x509 certificate secrets.

     * With parameters:
        * Application Name=sso
        * Custom RH-SSO Server Hostname=
        * JGroups Cluster Password=tpPjeJMQkWYTvtlmEo1A6mbAgR5oJRaw # generated
        * Datasource Minimum Pool Size=
        * Datasource Maximum Pool Size=
        * Datasource Transaction Isolation=
        * ImageStream Namespace=openshift
        * RH-SSO Administrator Username=MGOBwlID # generated
        * RH-SSO Administrator Password=OkqS07eehu1cvdOmAPr3s6EJrqBhOMqq # generated
        * RH-SSO Realm=
        * RH-SSO Service Username=
        * RH-SSO Service Password=
        * Container Memory Limit=1Gi

--> Creating resources ...
    configmap "sso-service-ca" created
    service "sso" created
    service "sso-ping" created
    route.route.openshift.io "sso" created
    deploymentconfig.apps.openshift.io "sso" created
--> Success
    Access your application via route 'sso-iankko-inject-ca-bundle.apps.rhsso43cluster.lab.rdu2.cee.redhat.com' 
    Run 'oc status' to view your app.
```

File: templates/openj9/sso74-openj9-x509-postgresql-persistent.json [hunk 2]
```diff
@@ -474,6 +493,12 @@
                                 "secret": {
                                     "secretName": "sso-x509-jgroups-secret"
                                 }
+                            },
+                            {
+                                "name": "service-ca",
+                                "configMap": {
+                                    "configMapName": "${APPLICATION_NAME}-service-ca"
```
Comment (iankko) [inline templates/openj9/sso74-openj9-x509-postgresql-persistent.json]: Ditto here (replace `"configMapName"` with plain `"name"`)

File: templates/sso74-x509-https.json [hunk 3]
```diff
@@ -369,6 +388,12 @@
                                 "secret": {
                                     "secretName": "sso-x509-jgroups-secret"
                                 }
+                            },
+                            {
+                                "name": "service-ca",
+                                "configMap": {
+                                    "configMapName": "${APPLICATION_NAME}-service-ca"
```
Comment (iankko) [inline templates/sso74-x509-https.json]: Yet here

File: templates/sso74-x509-postgresql-persistent.json [hunk 4]
```diff
@@ -474,6 +493,12 @@
                                 "secret": {
                                     "secretName": "sso-x509-jgroups-secret"
                                 }
+                            },
+                            {
+                                "name": "service-ca",
+                                "configMap": {
+                                    "configMapName": "${APPLICATION_NAME}-service-ca"
```
Comment (iankko) [inline templates/sso74-x509-postgresql-persistent.json]: And finally here

**Review by iankko — CHANGES_REQUESTED** [review]
Replace all instances of `"configMapName"` with plain `"name"` to get the templates working (tried on OCP v4.3.15 & it doesn't seem to understand / support the `"configMapName"` yet).

Once that is done, the template works OK (verified by testing)

**Comment by marun** [comment]
@iankko Ah, good catch. I modeled my change to the templates after the field name for mounting a secret (`secretName`) and forgot that configmaps are mounted with a LocalObjectReference that whose equivalent is `name`. Fixed!

File: templates/sso74-x509-postgresql-persistent.json (hunk 4 above)
Comment (marun) [reply templates/sso74-x509-postgresql-persistent.json]: Done

File: templates/sso74-x509-https.json (hunk 3 above)
Comment (marun) [reply templates/sso74-x509-https.json]: Done

File: templates/openj9/sso74-openj9-x509-postgresql-persistent.json (hunk 2 above)
Comment (marun) [reply templates/openj9/sso74-openj9-x509-postgresql-persistent.json]: Done

File: templates/openj9/sso74-openj9-x509-https.json (hunk 1 above)
Comment (marun) [reply templates/openj9/sso74-openj9-x509-https.json]: Done.

**Review by mhajas — CHANGES_REQUESTED** [review]
@marun Thanks for the PR! However, we need to somehow solve the testing of these. I see two options. 

- Either we can add additional test classes for added templates and run all tests on cluster we are currently using in our testing job
- Or, we can run different testing stages with prefixes `sso74-ocp4` and `sso74-openj9-ocp4`. This way we could easily change OCP versions on which templates are running. 

@marun Does this mean that starting OCP 4.5 the templates we currently have will stop working? How is it with support of the newer templates (*-ocp4-*) on OCP 3.x cluster? @iankko @hmlnarik We still support OCP 3.x right? 

WDYT? What approach would be better?

**Comment by marun** [comment]
> @marun Thanks for the PR! However, we need to somehow solve the testing of these. I see two options.
> 
> * Either we can add additional test classes for added templates and run all tests on cluster we are currently using in our testing job
> * Or, we can run different testing stages with prefixes `sso74-ocp4` and `sso74-openj9-ocp4`. This way we could easily change OCP versions on which templates are running.

The new templates have already been tested by @iankko as per https://github.com/jboss-container-images/redhat-sso-7-openshift-image/pull/145#pullrequestreview-416890031. My expectation from the email exchange is that updates to the tests can happen after my PRs merge. I'm not a java developer and I don't have access to whatever environment you're using for testing.

> @marun Does this mean that starting OCP 4.5 the templates we currently have will stop working? How is it with support of the newer templates (_-ocp4-_) on OCP 3.x cluster? @iankko @hmlnarik We still support OCP 3.x right?

The x509 templates without ocp4 in the name will be compatible with 3.x but not 4.x.

**Comment by hmlnarik** [comment]
Thanks @mhajas @marun!

> @marun Thanks for the PR! However, we need to somehow solve the testing of these. I see two options.
> 
> * Either we can add additional test classes for added templates and run all tests on cluster we are currently using in our testing job
> * Or, we can run different testing stages with prefixes `sso74-ocp4` and `sso74-openj9-ocp4`. This way we could easily change OCP versions on which templates are running.

The tests need to be addressed separately due to time constraints, so it is ok to leave it out to a separate PR by Keycloak team.

> @marun Does this mean that starting OCP 4.5 the templates we currently have will stop working? How is it with support of the newer templates (_-ocp4-_) on OCP 3.x cluster? @iankko @hmlnarik We still support OCP 3.x right?
> 
> WDYT? What approach would be better?

The -ocp4 templates are there only to support OCP 4 clusters, OCP 3 templates are still there for OCP3. The two are unfortunately incompatible, so both need to be maintained.

**Review by mhajas — APPROVED** [review]
I see. Thanks for the explanation. I am approving.

**Pull request merged**